README.md

```diff
@@ -1,44 +1,284 @@
+comparison operation applied to the current system call. It’s parametrised by
+the argument index, the length of the argument, the operator and the actual
+expected value.
+
+Going one step further, a `SeccompRule` is a vector of `SeccompCondition`s,
+that must all match for the rule to be considered matched. In other words, a
+rule is a collection of **and-bound** conditions for a system call.
+
+Finally, at the top level, there’s the `SeccompFilter`. The filter can be
+viewed as a collection of syscall-associated rules, with a predefined on-match
+action and a default action that is returned if none of the rules match.
+
+In a filter, each system call number maps to a vector of **or-bound** rules.
+In order for the filter to match, it is enough that one rule associated to the
+system call matches. A system call may also map to an empty rule vector, which
+means that the system call will match, regardless of the actual arguments.
+
+The following diagram models a simple filter, that only allows `accept4`,
+`fcntl(any, F_SETFD, FD_CLOEXEC, ..)` and `fcntl(any, F_GETFD, ...)`.
+For any other system calls, the process will be killed.
+
+![filter diagram](./docs/img/seccomp_filter.png)
+
+As specified earlier, there are two ways of expressing the filters:
+
+1. JSON (documented in [json_format.md](docs/json_format.md));
+2. Rust code (documented by the library).
+
+See below examples of both representation methods, for a filter equivalent to
+the diagram above:
+
+#### Example JSON filter
+
+```json
+{
+    "main_thread": {
+        "default_action": "kill_process",
+        "filter_action": "allow",
+        "filter": [
+            {
+                "syscall": "accept4"
+            },
+            {
+                "syscall": "fcntl",
+                "args": [
+                    {
+                        "index": 1,
+                        "type": "dword",
+                        "op": "eq",
+                        "val": 2,
+                        "comment": "F_SETFD"
+                    },
+                    {
+                        "index": 2,
+                        "type": "dword",
+                        "op": "eq",
+                        "val": 1,
+                        "comment": "FD_CLOEXEC"
+                    }
+                ]
+            },
+            {
+                "syscall": "fcntl",
+                "args": [
+                    {
+                        "index": 1,
+                        "type": "dword",
+                        "op": "eq",
+                        "val": 1,
+                        "comment": "F_GETFD"
+                    }
+                ]
+            }
+        ]
+    }
+}
+```
+
+Note that the JSON files need to specify a name for each filter. In the
+example above, there is only one: `main_thread`, other programs may be using
+multiple filters.
+
+#### Example Rust-based filter
+
+```rust
+SeccompFilter::new(
```

Comment:
Since we need to add comments about what parameter represents, maybe this is a good indicator that we should make the initialization easier.

When parameters have different types making mistakes in the initialization is hard, but in this particular case, both filter_action and default_action have the same type (SeccompAction). This makes it prone to errors as you can reverse the two without realizing.

The easiest way out is to use the builder pattern, but there might be other options worth exploring.

Reply:
I think we can handle this in the code review, and update the documentation afterwards.